modules/generator/processor/localblocks/processor.go

```diff
@@ -379,12 +435,25 @@ func (p *Processor) completeBlock() error {
 	p.blocksMtx.Lock()
 	defer p.blocksMtx.Unlock()
 
-	p.completeBlocks[(uuid.UUID)(newMeta.BlockID)] = ingester.NewLocalBlock(ctx, newBlock, p.wal.LocalBackend())
+	// Verify the WAL block still exists, it's possible that it went out of retention
+	// while it was being completed.
+	if _, ok := p.walBlocks[id]; !ok {
+		// WAL block is gone.
+		_ = level.Warn(p.logger).Log("msg", "local blocks processor WAL block disappeared while being completed, deleting complete block", "id", id)
+		err := p.wal.LocalBackend().ClearBlock(id, p.tenant)
+		if err != nil {
+			_ = level.Error(p.logger).Log("msg", "failed to clear complete block after WAL disappeared", "tenant", p.tenant, "block", id, "err", err)
+		}
+	}
+
+	p.completeBlocks[id] = ingester.NewLocalBlock(ctx, newBlock, p.wal.LocalBackend())
```

Comment:
Doesn't this add a phantom complete block if the WAL block has been deleted?

Reply:
Good catch, missing return statement when wal block has been deleted above.